### PR TITLE
Add texture in attribute color

### LINKED ISSUE #35591: Regression on the migrated Attributes pages
Tested on PR migration
- [x] B1: Wrong color input display
- [x] B2: Can't add attribute value from the list
- [ ] B3: Attribute value list doesn't adapt itself if the attribute type changes
- [x] B4: SQL query not working
- [ ] B5: Product doesn't adapt when the type of the attribute group changes
- [x] B6: Can't add texture This one is handled by https://github.com/PrestaShop/PrestaShop/pull/36068
- [x]  B7: Wrong wording titlte
- [x] B8: Missing wording and catch the error for the public name missing 
- [ ] B9: Can't click on the row to view the attribut value list
- [ ]  B10: Attribute group list is not by default sorted by position

I have found a few issues.
## Bug 1
The block Color used to displayed correctly. Now there's one line in the input instead of the chosen color.
**Before:**
Screenshot 2024-03-07 at 14 49 33
**After:**
Screenshot 2024-03-07 at 14 48 57

- [ ] Fixed ?

## Bug 2 

On the Attributes pages, I could chose to Add new attribute or Add a new value. Now I only have one choice.
**Before:**
Screenshot 2024-03-07 at 14 53 36
**After:**
Screenshot 2024-03-07 at 14 53 42

- [ ] Fixed ?

## Bug 3

When I add a new value and change the attribute group, the inputs for that attribute are not displayed. In my case, I have to go to my Color attribute and edit it to add a color manually.
Steps to reproduce :
- Go to BO > Attributes page > View `Paper Type`
- Click Add new value button
- In Attribute group > Change it to `Color`
- The inputs for the color attributes are not automatically displayed :x:

Same case the other way around.
- Go to BO > Attributes page > View `Color`
- Click Add new value button
- In Attribute group > Change it to `Dimension`
- The inputs for the color attributes are not automatically removed :x:

https://github.com/PrestaShop/PrestaShop/assets/16019289/e6524081-c975-4c07-bfb8-fd58b6b82b52

- [ ] Fixed ?
---

## Bug 4

The options `Show SQL Query` and `Export to SQL manager` do not work as expected. The query is different compared to the old page and the export gives a Bad Query.
**Before:**
Screenshot 2024-03-07 at 15 33 12
**After:**
Screenshot 2024-03-07 at 15 31 32

- [x] Fixed ?
---

## Bug 5

When I edit the Attribute Color to a Dropdown list Attribute type, it won't refresh my page with the right Attribute Type. On old page, it was well removed.
I tried clearing cache from PS and browser.

Steps to reproduce :
- Edit Color Attribute 
- Choose Attribute type = Drop-down list > Save
- Click View Color
- See all the Color column still displayed on the list

https://github.com/PrestaShop/PrestaShop/assets/16019289/9fe825e9-0cbc-44ce-bbf0-367e6c8e5f41

The issue is also present on the Product page. When I create a product with combinations, the attributes Color is displayed with a color, when I deleted it. (this last issue is not related to the PR)
Screenshot 2024-03-07 at 15 52 15

- [ ] Fixed ?

---

## Bug 6

The texture can't be added. The color Black is automatically added.

https://github.com/PrestaShop/PrestaShop/assets/16019289/2618d7b1-c277-4fec-bded-ece4faa7b14d

- [x] Fixed ?

---
## Bug 7

I have reproduced all of the issues mentioned by @Hlavtox https://github.com/PrestaShop/PrestaShop/pull/35270#pullrequestreview-1864601308
Indeed, when I edit an attribute value, the title of the page is misleading.
In my case, it says `Editing attribute Gray`. It should say `Editing attribute value Gray` or `Editing value Gray`.
There's no reference on the old page, as it only says "Attributes".

Screenshot 2024-03-07 at 15 08 14

- [x] Fixed ?
---
## Bug 8

If you save a new attribute without Public name, you will have an exception.

Screenshot 2024-03-07 at 15 18 39

- [x] Fixed ?
---
## Bug 9

Attribute list is not clickable, you have to click on the small View button to get to the list of values. On the old page, you could click on the name AND on the View button.

Same issue with Attributes values. you have to click on the Edit button to edit your value. On the old page, you could click on the name AND on the Edit button.

https://github.com/PrestaShop/PrestaShop/assets/16019289/a515741a-86ef-4e4f-a02a-2f157950d096

- [ ] Fixed ?

## Bug10

For some reason, when I open a list of attribute values, it automatically opens "rearrange" mode. It doesn't happen in attribute list.

Need to adapt the attribut group list

Screenshot 2024-03-07 at 15 58 35

- [ ] Fixed ?
--- 

### feedback 1
There's an option to Import your Attributes and attributes values. But in the import of PrestaShop, you have to import `Products` with Attributes in order to get the attributes created. So I think this is misleading and should not be an option. There's no Import type Attributes.

Screenshot 2024-03-07 at 15 24 53

### feedback 2

---

It is not possible to filter by Position anymore. https://github.com/PrestaShop/PrestaShop/issues/11217#issuecomment-661141039
**Before:**
Screenshot 2024-03-07 at 15 57 20

**After:**
Screenshot 2024-03-07 at 15 57 03

@MatShir comment: To order you have to select rearrange

### feedback 3

---

The block URL is misleading. It should not say URL. I'm not sure what it was meant to say 🤔 

Screenshot 2024-03-07 at 16 09 54

Regarding the block Meta Title, I'm not sure how to use it.
I added an element, but in the FO, I don't see it anywhere.

Screenshot 2024-03-07 at 16 13 38

@MatShir comment: You need to insert a path, maybe to check with the content team. It will be a bonus on this PR (not blocking) 

Waiting for your feedback 😃 
Thanks!

### PR DESCRIPTION
**Pull request by tleon**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | The texture box was not doing anyting and seemed no to be migrated from old page..
| Type?             | bug fix
| Category?         | BO 
| BC breaks?        |  no
| Deprecations?    |  no
| How to test?      | See #35591 bug 6
| UI Tests          | https://github.com/tleon/ga.tests.ui.pr/actions/runs/9117701983
| Fixed issue or discussion?     | Fixes #35591 
| Sponsor company   | PrestaShop SA

### CODE REVIEW

**Review by nicosomb — DISMISSED** [review]

File: src/Core/Form/IdentifiableObject/DataHandler/AttributeFormDataHandler.php
Comment (jolelievre) [inline src/Core/Form/IdentifiableObject/DataHandler/AttributeFormDataHandler.php]: This should be part of the command, you can inspire from what was done in other CQRS commands:
https://github.com/PrestaShop/PrestaShop/blob/fd1b5c16b7203bc418f31900a633c4370e9d5473/src/Core/Domain/Product/Image/Command/AddProductImageCommand.php
https://github.com/PrestaShop/PrestaShop/blob/3d06748975e00c548e940bf82862d9d211ae94d4/src/Core/Domain/Attachment/Command/AddAttachmentCommand.php

The commands should be passed with the path of the UploadFile (not an UploadFile instance, just a string), optionally you can also provide file size or mime type if needed

**Review by jolelievre — CHANGES_REQUESTED** [review]
Thanks @tleon 

all good for the grid part, the format part however shouldn't handle the upload in the controller, it must be the responsibility of the CQRS commands

File: src/Core/Form/IdentifiableObject/DataHandler/AttributeFormDataHandler.php
Comment (tleon) [reply src/Core/Form/IdentifiableObject/DataHandler/AttributeFormDataHandler.php]: Shouldn't the CQRS command reflect what is in the DB ? because the texture file path isn't in.  That's why I did not put it in the CQRS

**Review by nicosomb — DISMISSED** [review]

File: src/Adapter/File/Uploader/AttributeFileUploader.php
Comment (jolelievre) [inline src/Adapter/File/Uploader/AttributeFileUploader.php]:
```suggestion
    public function upload(string $filePath, int $id): void
```
I'm not sure null value is interesting to be authorized, this could be a valid string all the time, and it's the calling code that should check the value is not null before using this method

File: src/Adapter/File/Uploader/AttributeFileUploader.php
Comment (jolelievre) [inline src/Adapter/File/Uploader/AttributeFileUploader.php]:
```suggestion
            if (file_exists($filePath)) {
```
It can be interesting to check that the file exists

File: src/Adapter/Attribute/CommandHandler/AddAttributeHandler.php
Comment (jolelievre) [inline src/Adapter/Attribute/CommandHandler/AddAttributeHandler.php]: Since `$command->getPathName()` is optional and nullable it should be checked here `if (null !== $command->getPathName())`

File: src/Core/Domain/AttributeGroup/Attribute/Command/AddAttributeCommand.php
Comment (jolelievre) [inline src/Core/Domain/AttributeGroup/Attribute/Command/AddAttributeCommand.php]: The getter and the setter should have the same names for constencies, so either use `getTextureInformation/setTextureInformation` `getPathName/setPathName` or maybe `getTextureFilePath/getTextureFilePath`

File: src/Core/Domain/AttributeGroup/Attribute/Command/EditAttributeCommand.php
Comment (jolelievre) [inline src/Core/Domain/AttributeGroup/Attribute/Command/EditAttributeCommand.php]: SAme comment as in the add command

File: tests/Integration/Behaviour/Features/Context/Domain/AttributeFeatureContext.php
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/AttributeFeatureContext.php]:
```suggestion
        string $filePath = null
```
I believe this is never null

File: tests/Integration/Behaviour/Features/Scenario/Attribute/attribute_management.feature
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Attribute/attribute_management.feature]:
```suggestion
      | texture         | null             |
```

**Review by jolelievre — COMMENTED** [review]
Thanks @tleon 

A few comments here and there. However I have a question, do we need the possibility to remove a texture that was previously uploaded in the form? Because it seems like the command can handle the replacement of an image, but not the removal no?

File: src/Adapter/Attribute/CommandHandler/AddAttributeHandler.php
Comment (jolelievre) [inline src/Adapter/Attribute/CommandHandler/AddAttributeHandler.php]: It should be the uploader that checks if the file exists, the handler is only here to check if the data exists (if it's not null) Besides I'm not sure but maybe `file_exists` doesn't like null input?

**Review by jolelievre — COMMENTED** [review]
Thanks @tleon 

just one last comment about what the responsibility between the handler and the uploader service. And you didn't answer my question about about the removal of a texture is handled?

File: src/Adapter/File/Uploader/AttributeFileUploader.php
Comment (boherm) [inline src/Adapter/File/Uploader/AttributeFileUploader.php]: move_uploaded_file function doesn't throw exceptions and return boolean if it worked or not, but anyway for testing purpose, we can't change the try catch with a condition... So, is ok for now but we need to change all this type of lines!

File: src/Core/Grid/Data/Factory/AttributeGridDataFactory.php
Comment (boherm) [inline src/Core/Grid/Data/Factory/AttributeGridDataFactory.php]: Can we use properties promotion here?

File: src/PrestaShopBundle/Resources/config/services/adapter/attribute.yml
Comment (boherm) [inline src/PrestaShopBundle/Resources/config/services/adapter/attribute.yml:19]: I think this isn't necessary, no?

**Review by boherm — COMMENTED** [review]
LGTM, just few comments :)

File: src/Adapter/File/Uploader/AttributeFileUploader.php
Comment (jolelievre) [inline src/Adapter/File/Uploader/AttributeFileUploader.php:39]:
```suggestion
                move_uploaded_file($filePath, _PS_COL_IMG_DIR_ . $id . '.jpg');
```

File: src/Adapter/File/Uploader/AttributeFileUploader.php
Comment (jolelievre) [inline src/Adapter/File/Uploader/AttributeFileUploader.php:49]:
```suggestion
        if (file_exists(_PS_COL_IMG_DIR_ . $id . '.jpg')) {
            unlink(_PS_COL_IMG_DIR_ . $id . '.jpg');
```

File: src/Core/Grid/Data/Factory/AttributeGridDataFactory.php
Comment (jolelievre) [inline src/Core/Grid/Data/Factory/AttributeGridDataFactory.php:71]:
```suggestion
            if (file_exists(_PS_COL_IMG_DIR_ . (int) $record['id_attribute'] . '.jpg')) {
```

**Review by jolelievre — APPROVED** [review]
Just a suggestion to use existing const for this folder, not mandatory though

**Review by matks — APPROVED** [review]
LGTM

**Review by boherm — APPROVED** [review]

**Comment by paulnoelcholot** [comment]
Hello @tleon,

I tested your PR and it's OK for me! :tada: 

Thanks!

image

**Review by paulnoelcholot — APPROVED** [review]

**Comment by ps-jarvis** [comment]
QA approved, well done! Message to the maintainers: do not forget to milestone it before the merge.

**Pull request merged**
